### ISSUE TITLE
How to make agent run from bombs?

### ISSUE THREAD

**Issue by icaro56** [Unity-Technologies/ml-agents#644]
I am creating a example of Bomberman. I wish my agent ran from the bombs. So I add to my observations the relative bomb positions and the remaining timer to explode. But, my agent does not learn that the bomb can kill him and he is putting bomb and dying all the time and not try to run.

```
public override void CollectObservations()
    {
        Vector3 relativePosition = Target.position - this.transform.position;

        //posição relativa
        AddVectorObs(relativePosition.x / globalManager.xMax);
        AddVectorObs(relativePosition.z / globalManager.zMax);

        //distancia para cada borda da plataforma
        /*AddVectorObs((this.transform.position.x + globalManager.xMax) / globalManager.xMax);
        AddVectorObs((this.transform.position.x - globalManager.xMax) / globalManager.xMax);
        AddVectorObs((this.transform.position.z + globalManager.zMax) / globalManager.zMax);
        AddVectorObs((this.transform.position.z - globalManager.zMax) / globalManager.zMax);*/

        //velocidade do agente
        AddVectorObs(rigidBody.velocity.x / globalManager.xMax);
        AddVectorObs(rigidBody.velocity.z / globalManager.zMax);

        AddVectorObs(System.Convert.ToInt32(hasPlacedBomb));


        //colocar por enquanto apenas as 3 bombas mais próximas
        Vector3 agentPosition = myTransform.position;
        List<Bomb> bombsList = ServiceLocator.GetBombManager().getBombs(2);

        for (int i = 0; i < bombsList.Count; i++)
        {
            if (bombsList[i] != null)
            {
                //Debug.Log("Bombas: " + i + "  timer: " + bombsList[i].timer);
                Vector3 relativePositionToBomb = bombsList[i].transform.position - agentPosition;
                AddVectorObs(relativePositionToBomb.x / globalManager.xMax);
                AddVectorObs(relativePositionToBomb.z / globalManager.zMax);
                AddVectorObs(bombsList[i].timer / 3.0f);
            }
            else
            {
                Vector3 relativePositionToBomb = (new Vector3(9999999,1,99999999)) - agentPosition;
                AddVectorObs(relativePositionToBomb.x / globalManager.xMax);
                AddVectorObs(relativePositionToBomb.z / globalManager.zMax);
                AddVectorObs(1.0f);
            }
        }
    }

    private void penaltyNearbyBombs()
    {
        List<Bomb> bombsList = ServiceLocator.GetBombManager().getBombs(6);
        bool penalize = false;

        for (int i = 0; i < bombsList.Count; i++)
        {
            if (bombsList[i] != null && Vector3.Distance(myTransform.position, bombsList[i].transform.position) <= 4)
            {
                penalize = true;
                break;
            }
        }

        if (penalize)
        {
            AddReward(-1.0f);
        }
        else
        {
            AddReward(0.02f);
        }
    }

    public override void AgentAction(float[] vectorAction, string textAction)
    {
        hasPlacedBomb = false;
        bool putBomb = false;

        animator.SetBool("Walking", false);
        float vertical = Mathf.Clamp(vectorAction[0], -1, 1);
        float horizontal = Mathf.Clamp(vectorAction[1], -1, 1);
        float bombVal = Mathf.Clamp(vectorAction[2], 0, 1);
        putBomb = bombVal > 0.5f;
        bool stopped = (vertical > -0.5f && vertical < 0.5f) && (horizontal > -0.5f && horizontal < 0.5f);

        //recompensas
        float distanceToTarget = Vector3.Distance(this.transform.position, Target.position);

        //alcançou o objetivo
        if (distanceToTarget < 1.42f)
        {
            Done();
            AddReward(1.0f);
        }

        //se aproximando
        if (distanceToTarget < previousDistance)
        {
            AddReward(0.2f);
        }

        //Adicionar penalidade por estar próximo demais de uma bomba
        penaltyNearbyBombs();

        //penalidade de tempo
        AddReward(-0.05f);

        previousDistance = distanceToTarget;

        if (!stopped)
        {
            //cima
            if (vertical > 0.5f)
            {
                rigidBody.velocity = new Vector3(rigidBody.velocity.x, rigidBody.velocity.y, moveSpeed);
                //myTransform.rotation = Quaternion.Euler(0, 0, 0);
                animator.SetBool("Walking", true);
            } 
            //baixo
            else if (vertical < -0.5f)
            {
                rigidBody.velocity = new Vector3(rigidBody.velocity.x, rigidBody.velocity.y, -moveSpeed);
                //myTransform.rotation = Quaternion.Euler(0, 180, 0);
                animator.SetBool("Walking", true);
            }
            
            //direita
            if (horizontal > 0.5f)
            {
                rigidBody.velocity = new Vector3(moveSpeed, rigidBody.velocity.y, rigidBody.velocity.z);
                //myTransform.rotation = Quaternion.Euler(0, 90, 0);
                animator.SetBool("Walking", true);
            }
            //esquerda
            else if (horizontal < -0.5f)
            {
                rigidBody.velocity = new Vector3(-moveSpeed, rigidBody.velocity.y, rigidBody.velocity.z);
                //myTransform.rotation = Quaternion.Euler(0, 270, 0);
                animator.SetBool("Walking", true);
            }

            float dirX = this.transform.position.x - oldPosition.x;
            float dirZ = this.transform.position.z - oldPosition.z;

            if (Mathf.Abs(dirX) > Mathf.Abs(dirZ))
            {
                if (dirX >= 0)
                {
                    myTransform.rotation = Quaternion.Euler(0, 90, 0);
                }
                else
                {
                    myTransform.rotation = Quaternion.Euler(0, 270, 0);
                }
            }
            else
            {
                if (dirZ >= 0)
                {
                    myTransform.rotation = Quaternion.Euler(0, 0, 0);
                }
                else
                {
                    myTransform.rotation = Quaternion.Euler(0, 180, 0);
                }
            }
        }

        if (canDropBombs && putBomb)
        { //Drop bomb
            hasPlacedBomb = true;
            DropBomb();
        }

        oldPosition = this.transform.position;
    }
```

**Comment by vincentpierre** [comment]
Hi, here are a few things I see : 
 * Adding the distance to walls seems redundant, just giving `this.transform.position/ globalManager` should be enough I think.
 * I do not think the velocity and the position should be scaled the same way : `rigidBody.velocity.x / globalManager.xMaxVelocity`
 * I would set the agent to Done when it is killed by a bomb.
 * I think discrete control would be more appropriate given your scenario
 * I would try to simplify the problem to make sure the agent can solve simpler version first
 * I am not sure the animator will behave as expected with high simulation speed
 * I would try a curriculum in which the number of bombs slowly increases.

I hope this will help.

**Comment by icaro56** [comment]
Thanks @vincentpierre .

- The distance to walls is commented.
- Ok, I will change the velocity normalization
- So, I am setting to Done, but in the CollisionEvent
```
public void OnTriggerEnter (Collider other)
    {
        if (other.CompareTag ("Explosion"))
        {
            Debug.Log ("P" + playerNumber + " hit by explosion!");
            dead = true;
            AddReward(-1.0f);
            globalManager.PlayerDied(playerNumber);
            //Destroy(gameObject);

            Invoke("DoneWithDelay", .5f);
        }
    }

    private void DoneWithDelay()
    {
        Done();
    } 
```
-  Do you talk about discrete action space when speak about discrete control?
-  I do a simple version where the  agent need to achieve the goal (Any Transform), and it got.
-  The animator is little strange. But he's solving the problem for now.
-  Curriculum is interesting, but in this above example the agent has only one bomb.

I'm not getting the agent to learn that bombs kill.

**Comment by icaro56** [comment]
@vincentpierre , in the observation part that I get the bombs positions when there aren't bombs, which do I put in the observations? Currently, I am putting the diference between agent position and a position  far away.

```
List<Bomb> bombsList = ServiceLocator.GetBombManager().getBombs(2);

        for (int i = 0; i < bombsList.Count; i++)
        {
            if (bombsList[i] != null)
            {
                //Debug.Log("Bombas: " + i + "  timer: " + bombsList[i].timer);
                Vector3 relativePositionToBomb = bombsList[i].transform.position - agentPosition;
                AddVectorObs(relativePositionToBomb.x / globalManager.xMax);
                AddVectorObs(relativePositionToBomb.z / globalManager.zMax);
                AddVectorObs(bombsList[i].timer / 3.0f);
            }
            else
            {
                Vector3 relativePositionToBomb = (new Vector3(9999999,1,99999999)) - agentPosition;
                AddVectorObs(relativePositionToBomb.x / globalManager.xMax);
                AddVectorObs(relativePositionToBomb.z / globalManager.zMax);
                AddVectorObs(1.0f);
            }
        }
```

**Comment by icaro56** [comment]
I have done some modification. Now, the agent achieves the goal, but he still does not find that bombs is danger.

Now, I will modify the action space for discrete space like @vincentpierre suggested. The animation is very crazy, but it is better than before.

Thanks @vincentpierre 

I upload a video show the results on Youtube (https://youtu.be/593K045GvnQ)

**Comment by awjuliani** [comment]
Hi @icaro56 

The environment here looks nice. Since you are using a relatively fixed grid, one option could be to provide an encoding for each possible position in the grid about whether the bomb is present [1] or not [0]. You could do the same for the agent itself. This would make the observation size quite a bit larger, but would likely provide more interpretable information about the position of the bomb and agent relative to the platforms than their relative positions, which don't take into account the walls (which are critical in this problem).

**Comment by icaro56** [comment]
Thanks @awjuliani , Is there a example of fixed grid observation in the examples folder?

Let me see if I get it. Example: the observation 1 informs if there is bomb or not.

// GRID (0,0)
AddObservation (1)
// GRID (0,1)
AddObservation (0)
// GRID (0.2)
AddObservation (0)
// GRID (0.2)
AddObservation (0)

.
.
.

So I do this for all possible grid combinations? And I repeat more observations to inform if you have other agents or enemies for each combination of the grid?

Or I could tell that each grid can receive 0 = nothing, 0.1 = bomb, 0.2 = agent, 0.3 = etc, 0.4 = myself

**Comment by awjuliani** [comment]
@icaro56 Yep. That's what I have in mind. 

I wouldn't use partial values, since in the code you provided `bomb` would be semantically halfway between `nothing` and `agent` which doesn't really make sense. Separate "grids" would likely work better. Unfortunately we don't have an example exactly like this in the repo, but our "ray casting" system works similarly, except with an egocentric point of view, as opposed to a top-down view here.

**Comment by icaro56** [comment]
Thanks @awjuliani .

**Comment by icaro56** [comment]
@awjuliani , I thought something now. If do I create only one grid state? Where each important gameObject will be represented by a Integer. (e.g. Agent = 1, Block = 2, Bomb = 3, Goal = 4, None = 0)

1 | 0 | 0 | 0 | 0 | 0 | 0 | 
3 | 2 | 0 | 2 | 0 | 0 | 0 | 
0 | 2 | 0 | 2 | 0 | 0 | 0 | 
0 | 2 | 2 | 2 | 0 | 0 | 0 | 
0 | 2 | 0 | 0 | 0 | 0 | 0 | 
0 | 2 | 0 | 0 | 0 | 0 | 0 | 
0 | 2 | 0 | 0 | 0 | 0 | 0 | 
0 | 0 | 0 | 0 | 0 | 0 | 4 |

**Comment by awjuliani** [comment]
Hi @icaro56,

You will want to use a one-hot encoding instead of an integer. So instead of 1, 2, 3, you will want [1, 0, 0], [0, 1, 0], [0, 0, 1], etc.

**Comment by awjuliani** [comment]
Hi @icaro56,

That is correct, you would add multiple values per call to `AddVecObservation`. The function is designed to take variable length arrays as input. I would suggest looking at our RayCastPerception script for inspiration on adding one-hot vectors. https://github.com/Unity-Technologies/ml-agents/blob/master/unity-environment/Assets/ML-Agents/Examples/SharedAssets/Scripts/RayPerception.cs

**Comment by icaro56** [comment]
Thank you again.

**Comment by mattinjersey** [comment]
My comment in response to awjuliani. I think neural network can make filters, for example find if 0.2<x<0.4 . So I am not sure why it is invalid to set bomb=0.3  and player =0.1. Neural network should be able to figure this out and deal with it.
No?

**Comment by vincentpierre** [comment]
Using thresholds would make it unnecessarily harder for the network to "figure out". It might also require a bigger architecture to do what one-hot basically does. You can use this method to make it easier : 

```csharp
        protected void AddVectorObs(int observation, int range)
```

Where observation would be the int value and range be the maximum possible value. This method will convert the observation to a one hot.
One of the reasons using continuous values to encode discrete information is that it is a lot harder for the neural network to backpropagate since it assumes continuous values have an ordering. Similarly, you would not say for instance that `bomb > player` or that `bomb = player + 0.2`.

**Comment by icaro56** [comment]
I am using in my Bomberman experimente the One Hot Vector. But I can mix values. I named this new form of Hybrid vector. This way my observation vector does not grow as much as when using One Hot Vector.

One Hot Example:

Bomb = [0.1]
Agent = [1,0]
Bomb + Agent = [1,0,0]

Hybrid Example:

Bomb = [0.1]
Agent = [1,0]
Bomb + Agent = [1,1]

I did a lot of experiments, and actually the neural network learns much better with discretized values (0 or 1).
